tests/server/utils.py

```diff
@@ -1,154 +1,242 @@
+        verify: typing.Union[bool, str] = None,
+        cert: typing.Union[str, typing.Tuple[str, str]] = None,
+        json: typing.Any = None,  # pylint: disable=redefined-outer-name
+    ) -> Response:
+        if (
+            re.match(r"/?v[0-9](.[0-9]){0,2}/", url) is None
+            and not urlparse(url).scheme
+        ):
+            while url.startswith("/"):
+                url = url[1:]
+            url = f"/{self.version}{url}"
+        return super(OptimadeTestClient, self).request(
+            method=method,
+            url=url,
+            params=params,
+            data=data,
+            headers=headers,
+            cookies=cookies,
+            files=files,
+            auth=auth,
+            timeout=timeout,
+            allow_redirects=allow_redirects,
+            proxies=proxies,
+            hooks=hooks,
+            stream=stream,
+            verify=verify,
+            cert=cert,
+            json=json,
+        )
 
 
-class EndpointTestsMixin(SetClient):
-    """ Mixin "base" class for common tests between endpoints. """
+class BaseEndpointTests:
+    """Base class for common tests of endpoints"""
 
-    server: str = "regular"
     request_str: str = None
     response_cls: BaseModel = None
 
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
+    response: Response = None
+    json_response: dict = None
 
-        self.response = self.client.get(self.request_str)
-        self.json_response = self.response.json()
-        self.assertEqual(
-            self.response.status_code,
-            200,
-            msg=f"Request failed: {self.response.json()}",
+    @staticmethod
+    def check_keys(keys: list, response_subset: typing.Iterable):
+        for key in keys:
+            assert (
+                key in response_subset
+            ), f"{key} missing from response {response_subset}"
+
+    def test_response_okay(self):
+        """Make sure the response was successful"""
+        assert self.response.status_code == 200, (
+            f"Request to {self.request_str} failed: "
+            f"{json.dumps(self.json_response, indent=2)}"
         )
 
     def test_meta_response(self):
-        self.assertTrue("meta" in self.json_response)
-        meta_required_keys = [
-            "query",
-            "api_version",
+        """General test for `meta` property in response"""
+        assert "meta" in self.json_response
+        meta_required_keys = ResponseMeta.schema()["required"]
+        meta_optional_keys = list(
+            set(ResponseMeta.schema()["properties"].keys()) - set(meta_required_keys)
+        )
+        implemented_optional_keys = [
+            # "schema",
             "time_stamp",
             "data_returned",
-            "more_data_available",
             "provider",
+            "data_available",
+            "implementation",
         ]
-        meta_optional_keys = ["data_available", "implementation"]
+        # meta_may_keys = [
```

Comment:
Is there a reason these aren't being tested? Likewise schema above?

Reply:
Right. These needs to be included.
I keep forgetting the difference between server testing and validation tests... tsk tsk tsk.

Reply:
Ah. Yeah, so these meta values have not been implemented in the example server implementation. Hence they are commented out.
I will add a comment about this, keeping the commented out fields when/if we add them in the future.